### PR TITLE
Checkpoint 1 Review

### PR DESCRIPTION
**Pull request by femioladeji**
### Inverted Index

A front end application that allows for upload of only valid json files and indexes the words in the files with the documents that has those words. 
A search for any word in the file is done by searching the index and showing the documents in the file that has the word.
All uploaded files that has been indexed can also be searched

### CODE REVIEW

File: src/inverted-index.js [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+    // an empty object that keeps track of files uploaded
+    // and their index
+    this.filesIndexed = {};
+  }
+  
+
+  /**
+    createIndex function is used to get all the index
+    @param {object} jsonData- the json data to index
+    @parame {string} filename- the name of the file to be indexed
```
Comment (jadewale) [inline src/inverted-index.js]: **NITPICK** wrong spelling of param

File: src/inverted-index.js [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+    @return {boolean} true or false if the createIndex was successful
+  */
+  createIndex(jsonData, filename) {
+    //checks if the jsondata is accurate and not empty
+    try {
+      jsonData = JSON.parse(jsonData);
+    } catch(e) {
```
Comment (jadewale) [inline src/inverted-index.js]: **NITPICK** -- e can be changed to exception to make the variable name clearer

File: src/inverted-index.js [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+    //adds the filename as a key to the filesIndex and
+    //initializes it to an empty object
+    this.filesIndexed[filename] = {};
+    if(!this.prepareJsonData(jsonData, filename)) {
+      //if the file was not indexed, the key is removed
+      delete this.filesIndexed[filename];
+      return false;
+    }
+    return true;
+
```
Comment (jadewale) [inline src/inverted-index.js]: **NITPICK**-- unnecessary extra spacing. Not consistent with your other functions

File: src/inverted-index.js [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+  }
+
+  /**
+  prepareJsonData gets the json ready for indexing by tokenizing the
```
Comment (jadewale) [inline src/inverted-index.js]: **NITPICK** -- Comments should always be started with uppercase. I don't follow it sometimes

File: src/inverted-index.js [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+  */
+  createIndex(jsonData, filename) {
+    //checks if the jsondata is accurate and not empty
+    try {
+      jsonData = JSON.parse(jsonData);
+    } catch(e) {
+      return false;
+    }
+
+    //adds the filename as a key to the filesIndex and
```
Comment (jadewale) [inline src/inverted-index.js]: **NITPICK** -- Can be converted to a multi-line comment and started with uppercase

File: src/inverted-index.js [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+  }
+
+  /**
+  prepareJsonData gets the json ready for indexing by tokenizing the
+  statements
+  @param {object} jsonData - the jsonData that has been read from the file
+  @param {string} filename - the name of the file uploaded
+  @return {boolean} value to indicate if the index was successfully created
+  */
+  prepareJsonData(jsonData, filename) {
```
Comment (jadewale) [inline src/inverted-index.js]: **Code Quality** --prepareJsonData can be broken down into smaller functions. It is confusing to read

File: src/inverted-index.js [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+        if(indexDict[token].indexOf(documents[each].documentNum) === -1) {
+          indexDict[token].push(documents[each].documentNum);
+        }
+      }
+    }
+    return indexDict;
+  }
+
+  /**
+  getIndex method returns the indexed words and the documents they were found
```
Comment (jadewale) [inline src/inverted-index.js]: **NITPICK**-- correct the documents they were found to that were found.

File: src/inverted-index.js [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+  searchIndex searches the indexed words to determine the documents that the
+  searchterms can be found
+  @params searchTerm {string, array} the search query
+  @params filename {string}- the name of the file to search its index
+  @return {object|boolean} it returns boolean if the searchTerm is empty and
+  it retrns object if it is not. Each index is each searcykeyword.
+  Each with an array value of the document index
+  */
+  searchIndex(searchTerm, filename) {
+    if((typeof searchTerm === "string" && searchTerm.trim() === "") ||
```
Comment (jadewale) [inline src/inverted-index.js]: **NITPICK**-- Use of single quotes are in best practices for JS Strings.

File: src/inverted-index.js [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+  /**
+  searchIndex searches the indexed words to determine the documents that the
+  searchterms can be found
+  @params searchTerm {string, array} the search query
+  @params filename {string}- the name of the file to search its index
+  @return {object|boolean} it returns boolean if the searchTerm is empty and
+  it retrns object if it is not. Each index is each searcykeyword.
+  Each with an array value of the document index
+  */
+  searchIndex(searchTerm, filename) {
```
Comment (jadewale) [inline src/inverted-index.js:121]: **Code Quality** - function can be broken down into more functions

File: src/app.js [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+  let invIndex = new Index();
+  $scope.uploadedFiles = {};
+  $scope.allFlag = false;
+  $scope.allFilesIndexed = {};
+  /**
+  $scope.getIndex is executed when the getIndex button is clicked
+  */
+  $scope.getIndex = function() {
+    let fileChoice = $scope.uploadSelected;
+    if(fileChoice === undefined) {
```
Comment (jadewale) [inline src/app.js]: **Code Quality**  undefined is a falsy value, it always returns false. So if **!fileChoice** might be ideal here, only if your variable returns other falsy values and you are checking undefined explicitly.

File: src/app.js [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+          documents: invIndex.getDocuments(fileChoice),
+          indexedFile: fileChoice
+        }
+      ];
+      //keeps track of files that have been indexed
+      $scope.allFilesIndexed[fileChoice] = true;
+
+      //to check if two files or above have been indexed
+      //so that an option to search all files can be added
+      if(Object.keys($scope.allFilesIndexed).length > 1) {
```
Comment (jadewale) [inline src/app.js:40]: **Code quality** -- No need to check length > 1. if(obj.length) returns true if it is >= 1.

File: spec/inverted-index-test-spec.js [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+});
+
+describe("Search index", function() {
+  beforeEach(function() {
+    this.indexInstance = new Index();
+  });
+
+  it("should return an array of object(s) with each word as keys and the value is an array of the document index",
+    function() {
+      let book = '[{"title": "The hill","text": "Some may trust in"}, {"title": "Travis", "text": "The travis in CI is not in"}]';
```
Comment (sdowoade) [inline spec/inverted-index-test-spec.js]: Seems you don't have a `max-len` rule set for your linter.

File: src/app.js [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+  $scope.getIndex = function() {
+    let fileChoice = $scope.uploadSelected;
+    if(fileChoice === undefined) {
+      alert("Select a file to get index");
+      return false;
+    }
+    //if index was created for that file
+    if(invIndex.createIndex($scope.uploadedFiles[fileChoice].text, fileChoice)) {
+      //gets the indexed words
+      let indexes = invIndex.getIndex(fileChoice);
```
Comment (sdowoade) [inline src/app.js]: `const` is preferred if the variable(indexes) is not reassigned.

File: src/app.js [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+      //to check if two files or above have been indexed
+      //so that an option to search all files can be added
+      if(Object.keys($scope.allFilesIndexed).length > 1) {
+        $scope.allFlag = true;
+      }
+
+    } else {
+      // the file was not indexed because it is invalid;
+      delete $scope.uploadedFiles[fileChoice];
+      alert("Your json file is invalid, make sure each element has title and text property");
```
Comment (sdowoade) [inline src/app.js]: Minor thing. It's 2016 no one is big on alerts.

File: src/app.js [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+    if(!result) {
+      alert("Invalid search query");
+      return false;
+    }
+
+    $scope.indexed = result;
+    $scope.indexDisplay = false;
+  };
+
+  document.getElementById("uploadfile").addEventListener("change", readJson);
```
Comment (sdowoade) [inline src/app.js]: There might be a better way to do this. If you are using Angular stick to Angular conventions

File: src/inverted-index.js [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+
+  /**
+  getSearchResults method checks the index of the file and
+  returns the result
+  @param searchTokens {string or array} - the search query can be a string
+  or an array
+  @param filename {string} - the name of the file
+  @return {object} - an object with the found words as keys
+  */
+  getSearchResults(searchTokens, filename) {
```
Comment (sdowoade) [inline src/inverted-index.js:153]: This seems complex. Would be nice if you can simplify it. The Lodash/Underscore libraries might be useful.

File: src/inverted-index.js [hunk 17]
```diff
@@ -0,0 +1,200 @@
...
+    let indexToSearch = this.getIndex(filename), result = {};
+    //if it is a string of search terms then a split can be done
+    if(typeof searchTokens === "string") {
+      searchTokens = this.tokenize(searchTokens);
+    }
+    for(let indexCount in searchTokens) {
```
Comment (andela-dowoade) [inline src/inverted-index.js]: This easily has a complexity of O(n**2). It kinds of defeats the purpose of an Inverted Index.

Comment (andela-dowoade) [inline src/inverted-index.js]: There might be better a way to implement it.

File: src/inverted-index.js (hunk 16 above)
Comment (sdowoade) [reply src/inverted-index.js:153]: Also, consider the complexity of the function.

**Comment by makinwab** [comment]
@andela-foladeji you always want your commit messages to be consistent and follow a convention.

Check out this resource on writing good commit messages.
https://github.com/erlang/otp/wiki/writing-good-commit-messages

**Pull request merged**